<?xml version='1.0'?>
<sdf version='1.5'>
  <model name="morty">
    <static>false</static>
    <pose>0 0 0.5 0 0 0</pose>

    <link name="frame">
      <pose>0 0 0 0 0 0</pose>

      <inertial>
        <pose>-0.007506 0.000432 0.042077 0 0 0</pose>
        <mass>10.0</mass>
        <inertia>
          <ixx>0.07759446</ixx>
          <iyy>0.134953099</iyy>
          <izz>0.182993088</izz>
          <ixz>0.004729512</ixz>
          <ixy>-0.000038673</ixy>
          <iyz>0.000110587</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/frame-lod-3.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/frame-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <!-- Left side. -->
    <link name="rocker_axis_left">
      <pose>0 0.075 0.065 0 0 3.141592654</pose>

      <inertial>
        <pose>-0.000000 -0.047414 0.007917 0 0 0</pose>
        <mass>0.5</mass>
        <inertia>
          <ixx>0.000581951</ixx>
          <iyy>0.00039783</iyy>
          <izz>0.00055294</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.000036374</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-axis-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-axis-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <link name="rocker_left_front">
      <pose>-0.01 0.167 0.111 0 0 3.141592654</pose>

      <inertial>
        <pose>-0.103843 0.000207 -0.080442 0 0 0</pose>
        <mass>1.5</mass>
        <inertia>
          <ixx>0.0091575</ixx>
          <iyy>0.01351175</iyy>
          <izz>0.00480425</izz>
          <ixy>0.000048607</ixy>
          <ixz>-0.005074536</ixz>
          <iyz>-0.00003325</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-front-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-front-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <link name="rocker_left_rear">
      <pose>0.01 0.167 0.09 0 0 3.141592654</pose>

      <inertial>
        <pose>0.180409 -0.000496 -0.180775 0 0 0</pose>
        <mass>0.5</mass>
        <inertia>
          <ixx>0.004664009</ixx>
          <iyy>0.009240307</iyy>
          <izz>0.004668302</izz>
          <ixy>-0.000000849</ixy>
          <ixz>0.004392335</ixz>
          <iyz>0.000000165</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-rear-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-rear-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>


    <link name="bogie_left">
      <pose>0.128 0.167 -0.02625 0 0 0</pose>

      <inertial>
        <pose>-0.000000 -0.000283 -0.065016 0 0 0</pose>
        <mass>0.5</mass>
        <inertia>
          <ixx>0.001971824</ixx>
          <iyy>0.004614839</iyy>
          <izz>0.003279392</izz>
          <ixy>0.000000012</ixy>
          <ixz>0.0</ixz>
          <iyz>0.000009194</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/bogie-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/bogie-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <!-- Wheels -->
    <include>
      <uri>model://wheel_hhj30nt36v</uri>
      <pose>0.27477 0.14 -0.20783 1.570796327 0 0</pose>
      <name>wheel_left_front</name>
    </include>

    <include>
      <uri>model://wheel_hhj30nt36v</uri>
      <pose>-0.01596 0.14 -0.20783 1.570796327 0 0</pose>
      <name>wheel_left_middle</name>
    </include>

    <include>
      <uri>model://wheel_hhj30nt36v</uri>
      <pose>-0.30884 0.14 -0.20783 1.570796327 0 0</pose>
      <name>wheel_left_rear</name>
    </include>

    <!-- Joints. -->
    <joint type="revolute" name="left_frame_axis">
      <pose>0 0 0 0 0 0</pose>
      <parent>frame</parent>
      <child>rocker_axis_left</child>
      <axis>
        <xyz>0 1 0</xyz>
      </axis>
    </joint>
    <joint type="revolute" name="left_axis_rocker_rear">
      <pose>0 0 0 0 0 0</pose>
      <child>rocker_left_rear</child>
      <parent>rocker_axis_left</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="left_axis_rocker_front">
      <pose>0 0 0 0 0 0</pose>
      <child>rocker_left_front</child>
      <parent>rocker_axis_left</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="left_rocker_front_bogie">
      <pose>0 0 0 0 0 0</pose>
      <child>bogie_left</child>
      <parent>rocker_left_front</parent>
      <axis>
        <xyz>0 1 0</xyz>
        <limit>
          <lower>-0.785398163</lower>
          <upper>1.570796327</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="left_bogie_wheel_front">
      <pose>0 0 0.07 0 0 0</pose>
      <child>wheel_left_front::axis</child>
      <parent>bogie_left</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="left_bogie_wheel_middle">
      <pose>0 0 0.07 0 0 0</pose>
      <child>wheel_left_middle::axis</child>
      <parent>bogie_left</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="left_rocker_wheel_rear">
      <pose>0 0 0.07 0 0 0</pose>
      <child>wheel_left_rear::axis</child>
      <parent>rocker_left_rear</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>


    <!-- Right side. -->
    <!-- ########################################################################################################## -->
    <link name="rocker_axis_right">
      <pose>0 -0.075 0.065 0 0 0</pose>

      <inertial>
        <pose>-0.000000 -0.047414 0.007917 0 0 0</pose>
        <mass>0.5</mass>
        <inertia>
          <ixx>0.000581951</ixx>
          <iyy>0.00039783</iyy>
          <izz>0.00055294</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.000036374</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-axis-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-axis-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <link name="rocker_right_front">
      <pose>-0.01 -0.167 0.111 0 0 3.141592654</pose>

      <inertial>
        <pose>-0.103843 0.000207 -0.080442 0 0 0</pose>
        <mass>1.5</mass>
        <inertia>
          <ixx>0.0091575</ixx>
          <iyy>0.01351175</iyy>
          <izz>0.00480425</izz>
          <ixy>0.000048607</ixy>
          <ixz>-0.005074536</ixz>
          <iyz>-0.00003325</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-front-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-front-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <link name="rocker_right_rear">
      <pose>-0.01 -0.167 0.089 0 0 3.141592654</pose>

      <inertial>
        <pose>0.180409 -0.000496 -0.180775 0 0 0</pose>
        <mass>0.5</mass>
        <inertia>
          <ixx>0.004664009</ixx>
          <iyy>0.009240307</iyy>
          <izz>0.004668302</izz>
          <ixy>-0.000000849</ixy>
          <ixz>0.004392335</ixz>
          <iyz>0.000000165</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-rear-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/rocker-rear-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>


    <link name="bogie_right">
      <pose>0.128 -0.167 -0.02625 0 0 0</pose>

      <inertial>
        <pose>-0.000000 -0.000283 -0.065016 0 0 0</pose>
        <mass>0.5</mass>
        <inertia>
          <ixx>0.001971824</ixx>
          <iyy>0.004614839</iyy>
          <izz>0.003279392</izz>
          <ixy>0.000000012</ixy>
          <ixz>0.0</ixz>
          <iyz>0.000009194</iyz>
        </inertia>
      </inertial>

      <collision name="collision">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/bogie-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

      <visual name="visual">
        <geometry>
          <mesh>
            <uri>model://morty/meshes/bogie-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>
    </link>

    <!-- Wheels -->
    <include>
      <uri>model://wheel_hhj30nt36v</uri>
      <pose>0.27477 -0.14 -0.20783 -1.570796327 0 0</pose>
      <name>wheel_right_front</name>
    </include>

    <include>
      <uri>model://wheel_hhj30nt36v</uri>
      <pose>-0.01596 -0.14 -0.20783 -1.570796327 0 0</pose>
      <name>wheel_right_middle</name>
    </include>

    <include>
      <uri>model://wheel_hhj30nt36v</uri>
      <pose>-0.30884 -0.14 -0.20783 -1.570796327 0 0</pose>
      <name>wheel_right_rear</name>
    </include>

    <!-- Joints. -->
    <joint type="revolute" name="right_frame_axis">
      <pose>0 0 0 0 0 0</pose>
      <parent>frame</parent>
      <child>rocker_axis_right</child>
      <axis>
        <xyz>0 1 0</xyz>
      </axis>
    </joint>
    <joint type="revolute" name="right_axis_rocker_rear">
      <pose>0 0 0 0 0 0</pose>
      <child>rocker_right_rear</child>
      <parent>rocker_axis_right</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="right_axis_rocker_front">
      <pose>0 0 0 0 0 0</pose>
      <child>rocker_right_front</child>
      <parent>rocker_axis_right</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="right_rocker_front_bogie">
      <pose>0 0 0 0 0 0</pose>
      <child>bogie_right</child>
      <parent>rocker_right_front</parent>
      <axis>
        <xyz>0 1 0</xyz>
        <limit>
          <lower>-0.785398163</lower>
          <upper>1.570796327</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="right_bogie_wheel_front">
      <pose>0 0 0.07 0 0 0</pose>
      <child>wheel_right_front::axis</child>
      <parent>bogie_right</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="right_bogie_wheel_middle">
      <pose>0 0 0.07 0 0 0</pose>
      <child>wheel_right_middle::axis</child>
      <parent>bogie_right</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>
    <joint type="revolute" name="right_rocker_wheel_rear">
      <pose>0 0 0.07 0 0 0</pose>
      <child>wheel_right_rear::axis</child>
      <parent>rocker_right_rear</parent>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
        </limit>
      </axis>
    </joint>



    <!-- Diff -->
    <!-- ########################################################################################################## -->
    <!-- Sensors -->
    <link name="diff_connector">
      <pose>-0.2348023 0 0.1239466 0 0 0</pose>

      <inertial>
        <pose>-0.000000 -0.000001 0.000466 0 0 0</pose>
        <mass>0.1</mass>
        <inertia>
          <ixx>0.000574568</ixx>
          <iyy>0.000012108</iyy>
          <izz>0.000581622</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>-0.000000041</iyz>
        </inertia>
      </inertial>

      <visual name="visual">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/diff-connector-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/diff-connector-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

    </link>

    <link name="diff_bar_left">
      <pose>-0.1170713 0.115 0.1441726 0 0 0</pose>

      <inertial>
        <pose>-0.000022 -0.000077 -0.000075 0 0 0</pose>
        <mass>0.1</mass>
        <inertia>
          <ixx>0.000001273</ixx>
          <iyy>0.000770182</iyy>
          <izz>0.000770091</izz>
          <ixy>0.000000909</ixy>
          <ixz>-0.000000909</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>

      <visual name="visual">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/diff-bar-left-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/diff-bar-left-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>
    </link>

    <link name="diff_bar_right">
      <pose>-0.1170713 -0.115 0.1441726 0 0 0</pose>

      <inertial>
        <pose>-0.000022 -0.000077 -0.000075 0 0 0</pose>
        <mass>0.1</mass>
        <inertia>
          <ixx>0.000001273</ixx>
          <iyy>0.000770182</iyy>
          <izz>0.000770091</izz>
          <ixy>0.000000909</ixy>
          <ixz>-0.000000909</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>

      <visual name="visual">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/diff-bar-right-lod-2.dae</uri>
          </mesh>
        </geometry>
      </visual>

      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <mesh>
            <uri>model://morty/meshes/diff-bar-right-lod-1.dae</uri>
          </mesh>
        </geometry>
      </collision>

    </link>

    <joint type="revolute" name="frame_diff">
      <child>diff_connector</child>
      <parent>frame</parent>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint type="ball" name="diff_connector_bar_left">
      <pose>0.1168 0 0 0 0 0</pose>
      <child>diff_bar_left</child>
      <parent>diff_connector</parent>
    </joint>
    <joint type="ball" name="diff_bar_rocker_left">
      <pose>-0.1170713 0 0 0 0 0</pose>
      <child>diff_bar_left</child>
      <parent>rocker_axis_left</parent>
    </joint>
    <joint type="ball" name="diff_connector_bar_right">
      <pose>0.1168 0 0 0 0 0</pose>
      <child>diff_bar_right</child>
      <parent>diff_connector</parent>
    </joint>
    <joint type="ball" name="diff_bar_rocker_right">
      <pose>-0.1170713 0 0 0 0 0</pose>
      <child>diff_bar_right</child>
      <parent>rocker_axis_right</parent>
    </joint>


    <include>
      <uri>model://laser_leveling</uri>
      <pose>0.2 0 0 0 0 0</pose>
      <parent>frame</parent>
    </include>
    <include>
      <uri>model://rplidar_a2_m6</uri>
      <pose>0.226 0 0.085 0 0 0</pose>
      <parent>laser_leveling</parent>
    </include>
    <include>
      <uri>model://logitech_c920</uri>
      <pose>0.28 0 0 0 0 0</pose>
      <parent>frame</parent>
    </include>

    <!-- Joints for the sensors. -->
    <joint type="fixed" name="mount_leveling">
      <child>laser_leveling::dynamixel_roll</child>
      <parent>frame</parent>
    </joint>
    <joint type="fixed" name="laser_leveling">
      <child>rplidar_a2_m6::case</child>
      <parent>laser_leveling::mount_dynamixel_pitch</parent>
    </joint>
    <joint type="fixed" name="mount_camera">
      <child>logitech_c920::case</child>
      <parent>frame</parent>
    </joint>

    <plugin name="morty_motor_controller" filename="libgazebo-francor-morty-motor-controller-plugin.so">
      <motorLeftFrontJoint>wheel_left_front::motor</motorLeftFrontJoint>
      <motorLeftMidJoint>wheel_left_middle::motor</motorLeftMidJoint>
      <motorLeftRearJoint>wheel_left_rear::motor</motorLeftRearJoint>
      <motorRightFrontJoint>wheel_right_front::motor</motorRightFrontJoint>
      <motorRightMidJoint>wheel_right_middle::motor</motorRightMidJoint>
      <motorRightRearJoint>wheel_right_rear::motor</motorRightRearJoint>

      <motorLeftFrontPos>0.27477 0.17 0</motorLeftFrontPos>
      <motorLeftMidPos>0.01596 0.17 0</motorLeftMidPos>
      <motorLeftRearPos>-0.30884 0.17 0</motorLeftRearPos>
      <motorRightFrontPos>0.27477 -0.17 0</motorRightFrontPos>
      <motorRightMidPos>0.01596 -0.17 0</motorRightMidPos>
      <motorRightRearPos>-0.30884 -0.17 0</motorRightRearPos>

      <motorLeftFrontDiameter>0.22</motorLeftFrontDiameter>
      <motorLeftMidDiameter>0.22</motorLeftMidDiameter>
      <motorLeftRearDiameter>0.22</motorLeftRearDiameter>
      <motorRightFrontDiameter>0.22</motorRightFrontDiameter>
      <motorRightMidDiameter>0.22</motorRightMidDiameter>
      <motorRightRearDiameter>0.22</motorRightRearDiameter>

      <rosNamespace>morty</rosNamespace>
      <velocityTopic>velocity</velocityTopic>
      <odometryTopic>odom</odometryTopic>
    </plugin>

  </model>
</sdf>
